### PR TITLE
(PCP-637) Remove message persistence, multicast, and ActiveMQ

### PR DESCRIPTION
**Pull request by MikaelSmith**
Simplifies broker operation to lower memory use and complexity by
removing message persistence, delayed delivery, multicast delivery to
multiple targets, and the ActiveMQ library that was used for queueing
messages.

Message persistence adds significant complexity to PCP - tracking
message expiration/ttl, storing messages so they survive a restart -
without much benefit. It adds some durability around connection issues
on top of TCP/HTTP. However in practice, we require end-points to deal
with undeliverable messages and don't use expirations that significantly
differ from TCP delivery timeouts. Message persistence can be safely
removed without significantly impacting current use-cases, and should
improve the performance characteristics of the broker.

Multicast - the ability to use wildcards to address multiple recipients -
is another feature that adds implementation complexity, and in
practice is not used. Remove destination report as well, since it
no longer provides any useful information.

Both of these changes allow removal of the ActiveMQ component from
pcp-broker.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
I'm verifying this against pxp-agent acceptance tests, wait for confirmation they still pass.

**Comment by MikaelSmith** [comment]
Failed one test - it relied on the expiration time while the pxp-agent service restarted - so this will be blocked on tweaking a pxp-agent acceptance test.

**Comment by MikaelSmith** [comment]
The pxp-agent blocker is merged.

**Comment by MikaelSmith** [comment]
Rebased after Michal's changes.

File: test/integration/puppetlabs/pcp/broker/service_test.clj [hunk 1]
```diff
@@ -495,11 +467,10 @@
                           (message/set-expiry 5 :seconds)
                           (message/set-json-data "Hello"))]
           (client/send! client message)
-          (let [message (client/recv! client)]
-            (is (= "greeting" (:message_type message)))
-            (is (= "Hello" (message/get-json-data message)))))))))
```
Comment (mruzicka) [inline test/integration/puppetlabs/pcp/broker/service_test.clj]: Should we rather send a unicast message and keep these assertions? It seems that the assertions below only duplicate the `deliver-message-test` test.

File: src/puppetlabs/pcp/broker/core.clj [hunk 2]
```diff
@@ -127,158 +111,85 @@
...
+    data-content))
+
+(s/defn send-error-message
+  [in-reply-to-message :- (s/maybe Message) description :- s/Str connection :- Connection]
+  (let [data-content (make-error-data-content in-reply-to-message description)
+        error-msg (-> (message/make-message :message_type "http://puppetlabs.com/error_message"
+                                            :sender "pcp:///server")
+                      (message/set-json-data data-content))
+        error-msg (if in-reply-to-message
+                    (assoc error-msg :in-reply-to (:id in-reply-to-message))
```
Comment (MikaelSmith) [inline src/puppetlabs/pcp/broker/core.clj]: I think it's documented as `in_reply_to`. See https://tickets.puppetlabs.com/browse/PCP-312.

Comment (MikaelSmith) [inline src/puppetlabs/pcp/broker/core.clj]: I have a PR at https://github.com/puppetlabs/pcp-specifications/pull/38 updating the specification to declare `in_reply_to` usage. Currently it's usage in pcp-broker is unspecified in v1.0 (it was added to inventory response to tie it to a particular request).

File: src/puppetlabs/pcp/broker/core.clj [hunk 3]
```diff
@@ -127,158 +111,85 @@
...
 (s/defn deliver-message
-  "Message consumer. Delivers a message to the websocket indicated by the :target field"
-  [broker :- Broker capsule :- Capsule]
-  (if (and (:reject-expired-msg broker) (capsule/expired? capsule))
-    (process-expired-message broker capsule)
-    (if-let [websocket (get-websocket broker (:target capsule))]
+  "Message consumer. Delivers a message to the websocket indicated by the :targets field"
+  [broker :- Broker message :- Message sender :- (s/maybe Connection)]
+  ;; return error if multiple targets specified
+  (if (or (not= 1 (count (:targets message))) (p/uri-wildcard? (first (:targets message))))
```
Comment (MikaelSmith) [inline src/puppetlabs/pcp/broker/core.clj]: Yes, I'll change it.

**Comment by MikaelSmith** [comment]
Pushed code to address the other comments.

File: src/puppetlabs/pcp/broker/core.clj (hunk 2 above)
Comment (mruzicka) [inline src/puppetlabs/pcp/broker/core.clj]: Ah, ok.

**Pull request merged**